### ISSUE TITLE
Update docs to reflect new preferred `apply` method  on transform objects 

### ISSUE THREAD

**Issue by getdave** [WordPress/gutenberg#15972]
Please see 

A new `apply` method is now the preferred way to define the way in which one (or more blocks) can be transformed into one (or more) other blocks using the Block Transforms mechanic.

Previously developers were encouraged to define a  `transform` method on the [transform object definition](https://github.com/WordPress/gutenberg/blob/392bf02030ff4b25c478972911ca51ea21eef975/packages/block-library/src/cover/transforms.js#L16) in which they defined how to transform the block.

The issue is that this function is provided with only 2 arguments

1. Block `attributes`
2. Block `innerBlocks`

However, there are many cases where having access to the entire Block object would be required (eg: for example see [the Grouping mechanic](https://github.com/WordPress/gutenberg/pull/14908#discussion_r287433246)).

After a lengthy [discussion in Slack](https://wordpress.slack.com/archives/C02QB2JS7/p1559567845087000) and having reviewed several options, it was decided to create a new `apply` method which would accept the full block object. 

When the transform was applied then this `apply` method is preferred over any `transform` method. However, if it is not provided then the `transform` method is still run. As a result it is fully backwards compatible.

We now need to fully document this change. This involves

- [ ] Doing some typing

**Comment by getdave** [comment]
@youknowriad @aduth Did we decide to stick with this naming? 

Either way is this docs task _required_ for the release (which I think is coming up soon right?).

**Comment by aduth** [comment]
> @youknowriad @aduth Did we decide to stick with this naming?

I don't personally take strong issue with it. I know @nerrad had proposed an alternative `applyTransform` which is clear and respects the act of the application of a transform, but I find to be both verbose and somewhat redundant (with the `transforms` top-level property, not that the previous `transform` function wasn't already).

I think I'll be fine to see it go through as-is. We could flag it as "experimental" if we wanted some more time to consider it. But just as well, there is still an option to deprecate before the next WordPress stable release if a viable alternative presents itself.

> Either way is this docs task _required_ for the release (which I think is coming up soon right?).

No, I don't think it's required.

**Comment by aduth** [comment]
After talking a bit more with @nerrad , I think we may want to just mark this as experimental for now. There were a few other approaches we discussed, and in general I don't think we seem all that comfortable to commit to the current API as proposed. Since it's not documented anyways, we can plan to stabilize the API in tandem with introducing the documentation.

A few of the additional considerations:

- We could overload `to` and `from` to act as the complement to whichever is defined as the source array of block types.
   - For example, `{ from: [ 'core/paragraph' ], to( blocks ) { /* ... */ } }` or the inverse `{ to: [ 'core/paragraph' ], from( blocks ) { /* ... */ } }`
- To better clarify the expected return value, some more verbose alternative to describe the act of the transformation (e.g. `getTransformedBlocks`)

Personally, I also sought to look up additional options for words, closer to the intention with the "apply" as describe the _act_ of _applying a transform_. None of which really strike me more than what's already been considered.

- `execute`
- `run`
- `transact`
- `change`
- `act`
- `operate`
- `fulfill`

I've also considered completely different options for defining the transform, again not striking me as particularly viable or an improvement.

- A transform as a tuple of configuration and the function to carry it out (anonymous function)
- A transform as itself a function, either defining characteristics as properties of the function or in the return value. Possibly that all transform functions for a block type are run on any transform attempt, and a return value exists only if the block could carry it out.
- Revisiting the idea of the `transform` generator function idea as originally proposed in the discussion at https://github.com/WordPress/gutenberg/pull/14908#discussion_r287433246

As an immediate action item, I will follow-up with a pull request to mark this function as experimental.

**Comment by nerrad** [comment]
Thanks for summarizing this @aduth!

**Comment by getdave** [comment]
@aduth When might be remove `__exeperimental`? Is there a process for this?

**Comment by getdave** [comment]
@youknowriad @aduth Sorry to chase, but just wondering if it's ok to make this a first-class citizen now? I noticed [this effort to audit experimental APIs](https://github.com/WordPress/gutenberg/issues/17743) so perhaps this will be taken care of without me needing to keep 👁 👁 on it?

**Comment by aduth** [comment]
@getdave Sorry that I'd missed your October ping, as it happened to fall within a window of my extended leave from the project. Coincidentally, I'd been thinking about this since it had been [brought up in Slack](https://wordpress.slack.com/archives/C02QB2JS7/p1575490524144700) as something to stabilize. Much of this prior discussion had slipped my memory, and I don't personally take any issue with stabilizing the function as it exists today.

I can plan to take care of it, unless you had already been thinking about it?

**Comment by aduth** [comment]
Looking at the documentation:

https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/block-api/block-registration.md#transforms-optional

There are two things which stand out to me:

- Based on the previously-linked Slack discussion, we might want to consider the impact this has on the shortcode transform API, which is distinct from the block transform process, but adheres more to the API of the `transform` function (receives `attributes` as an argument). I suppose it depends whether we'd ever care to support the use-case of converting multiple shortcodes to one or more equivalent blocks (probably not), but in any case, there's now a bit more of a divergence in how the two are used.
- The usability of `convert` in transforming multiple blocks (always), vs. the previous usage of assuming to operate on a single block. This was obviously an intentional part of adding support for the multiple block conversion, but I consider that it could make the previous usage a bit more cumbersome, if the transform _did_ happen to be a one-to-one transformation. Ultimately, this would likely amount to an `Array#map` on the return value, but I don't know that this feels quite as nice as the previous behavior.

One idea which I don't know had been considered previously: What if we continue to support `transform`, and name the new function `transformMany`? Or some equivalent naming which implies that the intended usage is to convert multiple blocks. In this line of thinking, we don't need to advocate one or the other as being preferred, but each as equally sensible under certain conditions.

There's still a problem in that one form receives the full block objects (`transformMany( blocks )`) while the other only receives a subset of the block detail (`transform( attributes, innerBlocks )`). Again, this is where it might have just been preferable to have passed the `block` object to `transform` in the first place, but it's not something we can really change at this point in time.

What we *could* do, though, if we want the behavior to align, is to proceed with the renaming and soft-deprecation of `transform`, while still keeping the singular vs. many forms as two separate options.

Something like:

- `convert( block: WPBlock ): WPBlock|WPBlock[]`
- `convertMany( blocks: WPBlock[] ): WPBlock|WPBlock[]`

**Comment by aduth** [comment]
Looking again at the implementation, there's already an (undocumented) expectation that multi-block transformations should provide a `isMultiBlock: true` value as part of the transformation descriptor. Indeed, this is what the [group block already provides](https://github.com/WordPress/gutenberg/blob/772a5d5a018e0d1022a3ce837f9c38ca9a7ad805/packages/block-library/src/group/index.js#L89) to ensure that the conversion function receives an array of blocks.

Thus, I don't think we need to have separate functions, we can just name the function `convert`, and if the transformation also assigns `isMultiBlock: true`, then it will receive an argument as an array of blocks, otherwise it will receive as a single block.

**Comment by aduth** [comment]
> * Based on the previously-linked Slack discussion, we might want to consider the impact this has on the shortcode transform API, which is distinct from the block transform process, but adheres more to the API of the `transform` function (receives `attributes` as an argument). I suppose it depends whether we'd ever care to support the use-case of converting multiple shortcodes to one or more equivalent blocks (probably not), but in any case, there's now a bit more of a divergence in how the two are used.

Related to this, and probably more pressing for consideration: There are special kinds of transformation "types" which are handled outside the standard block transform API. These also use the `transform` naming, presumably as an overload of the block transformation `transform` function. I think for consistency's sake, we'd want to update these as well to prefer the consistent `convert` naming.

See also:

- `'prefix'` and `'files'` type transforms: https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/block-api/block-registration.md#transforms-optional
- Handlers:
  - https://github.com/WordPress/gutenberg/blob/772a5d5a018e0d1022a3ce837f9c38ca9a7ad805/packages/block-editor/src/components/block-drop-zone/index.js#L60-L74
  - https://github.com/WordPress/gutenberg/blob/772a5d5a018e0d1022a3ce837f9c38ca9a7ad805/packages/block-editor/src/components/rich-text/index.js#L282-L286

This late observation and additional maintenance overhead can serve as good justification that this ought to be refactored to a consolidated approach, which might be worth doing as part of this effort.

**Comment by aduth** [comment]
I managed to work my way through most of a branch to deprecate `transform` consistently in favor of `convert`, but encountered another issue in a potential inconsistency it introduces with the transforms `isMatch` function. The function receives an `attributes` object as its argument, much like the `transform` function. Granted, there's already a bit of inconsistency here in that it _does not_ receive `innerBlocks` as a second argument like the `transform` function had. We could decide to be content with this disparity. In retrospect, ideally this would receive a `block` or `blocks` object(s) just like an ideal `transform` function would have.

**Comment by aduth** [comment]
The pull request at #19401 will stabilize this API. The open questions are still pending on how we address some of the inconsistencies noted in my previous comment.

**Comment by nosolosw** [comment]
Cross-commenting for visibility https://github.com/WordPress/gutenberg/pull/19401#issuecomment-618983111

**Comment by nerrad** [comment]
I checked this against current trunk. This no longer reflects the API direction: #16047 deliberately downgraded convert to __experimentalConvert, #19401 never merged, and the current documentation correctly describes transform.

Closing as obsolete. If we want to stabilize convert in the future, that should start with a new, scoped API proposal.

See also the more recent #40208 which seems related in terms of better organization around transforms.

@getdave or @aduth if you disagree with this close it's fine to reopen.
